### PR TITLE
added collection to MOVE and MOVEXY

### PR DESCRIPTION
**Pull request by larkob**
This PR allows to use a collection as a source for MOVE and MOVEXY.

Note: Although you can define multiple `to` holders for MOVE, a collection will end up in the last holder. Technically, it first moves the collection to the first holder, then from the first to the next, and so on - so this is the correct (yet non-intuitive) behaviour.

At some later point we may want to allow distributing the collection across multiple holders, but that's beyond the scope of this PR because that functionality should have options for the desired distribution pattern.

Replaces PR #9 
-----------------------------------
To do:

- [ ] Final code check and conflict resolution since it is so old.
- [ ] Retest

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3373/pr-test (or any other room on that server)

**Review by 96LawDawg — APPROVED** [review]
Approved through test.  I also tested several games that use MOVE and MOVEXY and they still worked fine also.

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -900,6 +896,35 @@ export class Widget extends StateManaged {
     this.updatePiles();
   }
 
+  moveFlipToHolder(children, source, target, count, face) {
+    children.slice(0, count).reverse().forEach(c=> {
+      if(face !== null && c.flip)
+        c.flip(face);
+      if (!source && c.parent !== undefined) {
+        if (c.parent.parent !== undefined)
+          source = c.parent.parent;
```
[the remark this answers is not in the document]
Comment (larkob) [reply client/js/widgets/widget.js]: I've rewritten the code to clarify that this check tries to detect if widgets in the collection are already in the target holder.

**Review by 96LawDawg — CHANGES_REQUESTED** [review]
Doesn't work since the last time I checked.  In the test room, I tried a few things and it always just selected and moved one card.  The last thing I tried was a custom key ("movethiscard":true) and put it on 9 cards.  Debug shows the collection is correct, but only 1 card is moved.  Same problem with MOVEXY.

**Comment by bjalder26** [comment]
> Doesn't work since the last time I checked. In the test room, I tried a few things and it always just selected and moved one card. The last thing I tried was a custom key ("movethiscard":true) and put it on 9 cards. Debug shows the collection is correct, but only 1 card is moved. Same problem with MOVEXY.

I believe that was because it defaults to a count of 1.  I changed it to a high number, and it appears to work.  Likewise, lower numbers appear to work as well.

**Comment by bjalder26** [comment]
I noticed that when you select cards to `move` using a given property that if the count is limited it will not keep moving NEW cards into the holder, but instead will keep moving the same cards into the holder.  Maybe this is the desired behavior as the property can be removed or altered with `onEnter` (which is what I did) or through a `SET`.  However, I wanted to point it out in case you had something else in mind for how it was supposed to work.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.  Interitus pointed out the issue I had was failing to use the `count` parameter.  With the clarification, this works as expected and can be merged.  Unfortunately, due to passage of time, it has gotten out of synch with the main branch and will require some hand's on tweaks.

**Comment by robartsd** [comment]
Conflict in jsonedit.js is easy to resolve; conflict resolution is just picking up the removal of RANDOM adjacent to these changes.

Conflict in widget.js is more challenging. The last conflict here is simply that the new move helper functions are directly adjacent to a function that had async added. The other two conflicts involve the actual functions this changed having changes from main as well; most likely the new code needs `async` and `await` added appropriately (possibly also for the helper functions).

**Comment by bjalder26** [comment]
It looks like the first conflict here is that move and moveXY were are both updated in jsonedits.js AND that RANDOM was deleted in #488.  S it looks like we need to keep the move and moveXY from this PR, but delete RANDOM. https://github.com/ArnoldSmith86/virtualtabletop/pull/488/files
image

**Comment by bjalder26** [comment]
It looks like the 2nd conflict is centered around a change from `foreach` to `for of` and the addition of awaits and async. Those changes were made as part of #406.
image

**Review by rogerl50 — CHANGES_REQUESTED** [review]
Please add a comment containing draft language to update the Wiki when this PR is merged, so we can keep the Wiki up to date.

**Comment by bjalder26** [comment]
> Please add a comment containing draft language to update the Wiki when this PR is merged, so we can keep the Wiki up to date.

I would think something like this:

## MOVE

This function moves widgets into a target holder.

Parameters:
* from: _holderID_ (or an array) - specifies the holder(s) that contains the widgets to move.
* collection: _collection name_ - specifies the collection of widgets to move.
* to: _holderID_ (or an array) - specifies the holder(s) that widgets should be moved into.  Since movement into only 1 holder is supported at this time, widgets will be moved into the last holder in an array.
* count: _number_ - limits the amount of moved widgets (defaults to 1 if omitted; use 0 to indicate no limit)

The parameter _face_ can optionally set the face of the moved widgets to the given value (see FLIP). If omitted, the widgets will be left as they are.

```JS
{
  "func": "MOVE",
  "from": "holderID",
  "to": "holderID",
  "count": 0
}
```
```JS
{
  "func": "MOVE",
  "collection": "myCollection",
  "to": "holderID",
  "count": 0
}
```

## MOVEXY

This function moves widgets to a specific position on the surface (i.e. outside any holders). In case the widgets are currently contained in another widget (e.g. pile or holder), they will be moved out before being moved to the given coordinates.

Parameters:
* from: _holderID_ (or an array) - specifies the holder that contains the widgets to move
* collection: _collection name_ - specifies the collection of widgets to move.
* count: _number_ - limits the amount of moved widgets (defaults to 1 if omitted; use 0 to indicate no limit)
* x: _number_ - the x position on the surface (defaults to 0)
* y: _number_ - the y position on the surface (defaults to 0)

The parameter _z_ can optionally set the z-order of the widget. If omitted, the z-order will be determined automatically.

The parameter _face_ can optionally set the face of the moved widgets to the given value (see FLIP). If omitted, the widgets will be left as they are.

Please note that the moved widgets can form _piles_ as if they were moved by players. They will NOT be placed into any _holders_ though, even if the coordinates match perfectly. Use _MOVE_ for that.

```JS
{
  "func": "MOVEXY",
  "from": "holderID",
  "count": 2,
  "face": 0,
  "x": 100,
  "y": 100,
  "z": 100
}
```
```JS
{
  "func": "MOVEXY",
  "collection": "myCollection",
  "count": 10,
  "face": 1,
  "x": 200,
  "y": 200,
  "z": 100
}
```

**Review by bjalder26 — APPROVED** [review]
Still seems to work after the conflicts were resolved.  Approved through testing.

**Comment by ArnoldSmith86** [comment]
Yeah. I'll just look at the code today and make another review.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -1180,6 +1171,36 @@ export class Widget extends StateManaged {
...
   }
 
+  async moveFlipToHolder(children, source, target, count, face) {
+    for (const c of children.slice(0, count || 999999).reverse()) {
+      if(face !== null && c.flip)
+        await c.flip(face);
+      let isInTargetHolder = (source == target);
+      if (!source && c.parent !== undefined) {
+        if (c.parent.get('type') == 'pile' && c.parent.parent !== undefined)
+          isInTargetHolder = (c.parent.parent == target);
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:1181]: You shouldn't use `c.parent` here. I'm pretty sure it's not what you want. It's not always the same as `widgets.get(c.get('parent'))` which is probably what you need here.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I think it would be more consistent if a `count: 5` to multiple target holders would move 5 cards into each holder.

**Comment by bjalder26** [comment]
The 'conflict' here is from the addition of these lines in green.  The real conflict was already resolved.

image

**Comment by rogerl50** [comment]
This seems like a useful enhancement, and shouldn't require too much work to get it fixed and merged. There are conflicts that need to be dealt with (not only the trivial issue mentioned above, but a new one brought about by the introduction of anonymous collections), and a couple of comments above that need to be resolved.

**Comment by 96LawDawg** [comment]
The only thing remaining from #60 are MOVE and MOVEXY.  All the other applicable functions have been modified to work with collections.  Seems like we should get this done for consistency.

**Comment by ArnoldSmith86** [comment]
Superseded by #1658 and #1888.

**Pull request closed without merging**